File: api/src/routes/hdw/statistics.js
/* eslint-disable comment-length/limit-single-line-comments */
const express = require('express');
// const { query } = require('express-validator');
// const _ = require('lodash/fp');

const prisma = require('@/db');
const asyncHandler = require('@/middleware/asyncHandler');
// const { validate } = require('@/middleware/validators');
// const { accessControl } = require('@/middleware/auth');

const router = express.Router();
// const isPermittedTo = accessControl('cohorts');

router.get('/', asyncHandler(async (req, res) => {
  // #swagger.operationId = 'statistics'
  // #swagger.tags = ['statistics']
  // #swagger.summary = 'Get statistics'
  // #swagger.description = 'Returns various statistics about the database'

  const rows = await prisma.$queryRaw`
    select
      (select reltuples FROM pg_class WHERE relname = 'participant') as subjects,
      (select reltuples FROM pg_class WHERE relname = 'dx' ) as diagnoses,
      (select reltuples FROM pg_class WHERE relname = 'procedure') as procedures,
      (select reltuples FROM pg_class WHERE relname = 'encounter') as encounters
  `;

  res.json(rows[0] || {
    subjects: 0,
    diagnoses: 0,
    procedures: 0,
    encounters: 0,
  });
}));

module.exports = router;

File: python_api/api/routers/statistics.py
from fastapi import APIRouter

from api.db import get_connection


def stats() -> dict:
    conn = get_connection()
    with conn:
        cursor = conn.cursor()
        sql = """
    select
      (select count(*) from subject s) as subjects,
      (select count(*) from "dx" d) as diagnoses,
      (select count(*) from "procedure" p) as procedures,
      (select count(*) from encounter e) as encounters
    """
        cursor.execute(sql)
        return dict(cursor.fetchone())


router = APIRouter(
    prefix="/statistics",
    tags=["statistics"],
)


@router.get("/")
def get_statistics():
    return stats()
